libcudacxx/include/cuda/__floating_point/operators.h

```diff
@@ -0,0 +1,1250 @@
+
+  if constexpr (__rank_order == __fp_conv_rank_order::__equal)
+  {
+    // Extended floating point types have higher subrank, prefer cuda extended types over std extended types
+    // Fixme: potentially will not work correctly for long double
+    // auto val = 1.0f64 + 1.0l; // val will be of type long double, is this right?
+    if constexpr (__is_standard_floating_point_v<_Lhs> || __is_std_extended_floating_point_v<_Lhs>)
+    {
+      return _Rhs{};
+    }
+    else
+    {
+      return _Lhs{};
+    }
+  }
+  else if constexpr (__rank_order == __fp_conv_rank_order::__greater)
+  {
+    return _Lhs{};
+  }
+  else if constexpr (__rank_order == __fp_conv_rank_order::__less)
+  {
+    return _Rhs{};
+  }
+  else
+  {
+    static_assert(__always_false<sizeof(_Lhs)>(), "Cannot make a common fp type from the given types");
+    _CCCL_UNREACHABLE();
+  }
+}
+
+template <class _Lhs, typename _Rhs>
+using __fp_common_type_t = decltype(__fp_make_common_type<_Lhs, _Rhs>());
+
+// Implementations of the arithmetic operations. Usually an operation is in several parts:
+//   1. __op() - the entry point for the operation which tries to implement the operation using the host & device native
+//      types if available, otherwise calls the __op_impl() function.
+//
+//   2. __op_impl() - the non native implementation of the operation for the given type. This function dispatches to the
+//      target specific implementation and falls back to the constexpr implementation.
+//
+//   3. __op_impl_TARGET() - the target specific implementation. If available, implements the operation via asm.
+//
+//   4. __op_impl_constexpr() - the constexpr implementation of the operation. Slow.
+//
+// The arguments may be heterogenous. In that case the implementation is chosen in the following order:
+//   1. try to use the host & device native types
+//   2. try to use host & device native instructions for mixed arithmetic
+//   3. cast the arguments to the common type and use the implementation for homogeneous arguments
+struct __fp_ops
+{
+  /********************************************************************************************************************/
+  // Negation
+  /********************************************************************************************************************/
+  template <class _Tp>
+  _CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI constexpr static _Tp __neg_impl_constexpr(const _Tp& __src) noexcept
+  {
+    auto __ret{__src};
+    __ret.__set_sign(!__ret.__get_sign());
+    return __ret;
+  }
+
+  template <class _Tp>
+  _CCCL_NODISCARD _CCCL_HIDE_FROM_ABI _CCCL_DEVICE static _Tp __neg_impl_device(const _Tp& __src) noexcept
+  {
+    [[maybe_unused]] _Tp __ret;
+
+#  if __cccl_ptx_isa >= 600
+    if constexpr (_CUDA_VSTD::_CCCL_TRAIT(is_same, _Tp, fp16))
+    {
+      NV_IF_TARGET(NV_PROVIDES_SM_53,
+                   (asm("neg.f16 %0, %1;" : "=h"(__ret.__storage_) : "h"(__src.__storage_)); return __ret;))
```

Comment:
Once again - this really should go into codegen. And you can use a "trick" of having the default implementation (`__neg_impl_constexpr`, in this case) and the specialized implementations (containing the inline asm) be in the same overload set, with the specialized ones being better matches. This way, a lot of the inline logic goes away (turning the selection of the implementation into a much more declarative thing).